### PR TITLE
fix(eslint-plugin): [no-misused-promises] avoid unnecessary calls to getContextualType

### LINKED ISSUE #6186: Enhancement: no-misused-promises small performance improvement, avoiding unnecessary call to getContextualType
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

I profiled my eslint to see what it is slow, and it lead me to `@typescript-eslint/no-misused-promises`, and the culprate inside is `getContextualType`
it appears that `returnsThenable` can be called first before `getContextualType` for `checkJSXAttribute`

profiling a smaller component:
image

https://github.com/typescript-eslint/typescript-eslint/blob/v5.46.0/packages/eslint-plugin/src/rules/no-misused-promises.ts#L378-L387

It looks like `getContextualType` is more expensive than `returnsThenable`

Inside I saw that there was potential for an optimization to not call `checker.getContextualType(value)` every time. I made this change locally and my lints on my big project went from ~130 seconds to ~125

There appears to be some other optimizations that can be done to calls to `getContextualType` but I only looked at 1 place, and I'm not sure of the technicals of `getContextualType` and `returnsThenable` yet

on my large private repo, I got 4311 short circuits and 15 calls to `getContextualType`. so it saved `4311` calls to `getContextualType`.

I am willing to make the PR

```diff
diff --git a/packages/eslint-plugin/src/rules/no-misused-promises.ts b/packages/eslint-plugin/src/rules/no-misused-promises.ts
index b6914ae2..0e85656e 100644
--- a/packages/eslint-plugin/src/rules/no-misused-promises.ts
+++ b/packages/eslint-plugin/src/rules/no-misused-promises.ts
@@ -375,11 +375,16 @@ export default util.createRule<Options, MessageId>({
       ) {
         return;
       }
+      if(!returnsThenable(checker, value.expression)){
+        return;
+      }
       const contextualType = checker.getContextualType(value);
       if (
         contextualType !== undefined &&
-        isVoidReturningFunctionType(checker, value, contextualType) &&
-        returnsThenable(checker, value.expression)
+        isVoidReturningFunctionType(checker, value, contextualType)
       ) {
         context.report({
           messageId: 'voidReturnAttribute',
```

### Reproduction Repository Link

N/A

### Repro Steps

1) clone the repo
2) `yarn install`
3) `yarn lint`
4) run `NODE_OPTIONS="--inspect-brk" ` prefixing the eslint command
5) observe the slowest inside the chrome profiler

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.45.1` |
| `@typescript-eslint/parser`            | `5.45.1` |
| `TypeScript`                           | `4.9.3` |
| `ESLint`                               | `8.29.0` |
| `node`                                 | `v16.17.0` |

### PR DESCRIPTION
**Pull request by scottarver**
## PR Checklist

- [x] Addresses an existing open issue: fixes #6186
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

It appears that `returnsThenable` is faster than `getContextualType` in almost all cases. 
This PR only calls `getContextualType` when needed, after `returnsThenable` returns true.

I ran numbers on how often this is the case: Out of 19,425 total checks, only 96 (33+63) needed a call to  `getContextualType`, this saved 19329 (17289 + 2040 )  calls to `getContextualType` by short circuiting with `returnsThenable`
Also of note on average, `returnsThenable` took 0.12891074558698418ms to run vs `getContextualType` taking 0.5058407855855643ms, saving 9.8 seconds. 
```
total time saved:  9817.560766001232
{
  thenableAndContextual: 33,
  NotThenableAndContextual: 17289,
  thenableAndNotContextual: 63,
  NotThenableAndNotContextual: 2040
}
thenable count:  19425
thenable average: 0.12891074558698418
contextualType count:  19425
contextualType average: 0.5058407855855643
```

Using the chrome profiler to compare:  ran with `node --inspect-brk ./node_modules/.bin/eslint src`
Each check* fn has to be looked at together as a whole, in the table below on a single run it saved 1.2 seconds
| Before | After |
| ------------- | ------------- |
| image | image |

  | before | after |   |   |  
-- | -- | -- | -- | -- | --
checkTestConditional | 73.6 | 77 |   |   |  
checkConditional | 2.4 | 3.6 |   |   |  
checkAssignment | 1.7 | 2.9 |   |   |  
checkVariableDeclaration | 1160.5 | 1207.4 |   |   |  
checkProperty | 74.4 | 40.2 |   |   |  
checkReturnStatement | 9.2 | 56.1 |   |   |  
checkJSXAttribute | 2076.9 | 754.7 |   |   |  
checkArguments | 121.6 | 125.6 |   |   |  
checkSpread | 5.5 | 4.4 |   | ms diff | seconds
  | 3525.8 | 2271.9 |   | 1253.9 | 1.2539

### CODE REVIEW

File: packages/eslint-plugin/src/rules/no-misused-promises.ts [hunk 1]
```diff
@@ -269,27 +269,31 @@ export default util.createRule<Options, MessageId>({
...
+          )
         ) {
           context.report({
             messageId: 'voidReturnProperty',
             node: node.value,
           });
         }
       } else if (ts.isShorthandPropertyAssignment(tsNode)) {
+        if (!returnsThenable(checker, tsNode.name)) {
+          return;
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-misused-promises.ts:291]: > Added line #L291 was not covered by tests

If these lines aren't getting hit by tests, then one of two things is possible:

* These lines are unhittable and can be removed
* The tests are missing cases that your code does hit

Do you have examples of the latter? If not, I'd assume it's the former and we can remove the lines.

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
This started out as a nice performance improvement - and now is also an unnecessary code removal and/or test expansion opportunity! Great either way 😄

**Comment by JoshuaKGoldberg** [comment]
👋 ping @scottarver - is this still something you have time for? No worries if not, just checking in.

**Comment by JoshuaKGoldberg** [comment]
Ooh, I removed some more calls and all of the unit tests are failing... but now we've got lint failures in the project itself. So we've got to add unit tests! 🙌 

https://cloud.nx.app/runs/NHi3D9Xh0H

```plaintext
/home/runner/work/typescript-eslint/typescript-eslint/packages/eslint-plugin/src/rules/adjacent-overload-signatures.ts
  172:18  error  Promise-returning function provided to property where a void return was expected  @typescript-eslint/no-misused-promises
  173:16  error  Promise-returning function provided to property where a void return was expected  @typescript-eslint/no-misused-promises
  174:22  error  Promise-returning function provided to property where a void return was expected  @typescript-eslint/no-misused-promises
  175:22  error  Promise-returning function provided to property where a void return was expected  @typescript-eslint/no-misused-promises
  176:24  error  Promise-returning function provided to property where a void return was expected  @typescript-eslint/no-misused-promises
  177:23  error  Promise-returning function provided to property where a void return was expected  @typescript-eslint/no-misused-promises

/home/runner/work/typescript-eslint/typescript-eslint/packages/eslint-plugin/src/rules/ban-types.ts
  203:41  error  Promise-returning function provided to variable where a void return was expected  @typescript-eslint/no-misused-promises

/home/runner/work/typescript-eslint/typescript-eslint/packages/eslint-plugin/src/rules/block-spacing.ts
...
```

**Comment by jakebailey** [comment]
Truthfully I didn't even realize these APIs were available, let alone useful externally. Clearly, I'm wrong.

I think someone more experienced than me will have to write out explanations of those. I'll file an issue to track this request; meant to do that post-meeting but forgot.

**Pull request merged**